### PR TITLE
Improved Diagnostic for C-style for loop

### PR DESCRIPTION
**Pull request by CippoX**
Improved Diagnostic for C-style for loop using `getTokens` introduced by #1464

### CODE REVIEW

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by CippoX** [comment]
I have to think about the right approach

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift [hunk 1]
```diff
@@ -662,33 +662,21 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
...
-          Syntax(node.unexpectedBetweenTypeAnnotationAndInKeyword),
-          Syntax(node.inKeyword),
-          Syntax(node.unexpectedBetweenInKeywordAndSequenceExpr),
-          Syntax(node.sequenceExpr),
-          Syntax(node.unexpectedBetweenSequenceExprAndWhereClause),
-          Syntax(node.whereClause),
-          Syntax(node.unexpectedBetweenWhereClauseAndBody),
-          Syntax(unexpectedCondition),
-        ] as [Syntax?]).compactMap({ $0 }),
+        highlights: (getTokens(between: firstToken, and: lastToken).map { Optional(Syntax($0)) }).compactMap({ $0 }),
```
Comment (ahoppen) [inline Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:672]: Why do you need to explicitly spell out `Optional` here?

Comment (CippoX) [reply Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:672]: I was thinking about replicating `as [Syntax?]`,
However, the test failing is `testColoringWithHighlights()`. The problem seems to be that with `getTokens` we are creating an array of tokens, which is not what was happening previously. Do you think I should find a way to walk through `node`'s children?

**Comment by CippoX** [comment]
What about this solution? Obviously, it's very primitive, if you think that's the right approach I'll refine it

**Comment by ahoppen** [comment]
If we need to write a new function for this, I don’t think it’s worth it. The C-style for loop diagnostic is a bit of a weirdo anyway and unless a new use case for `getNodes(between:and:)` comes up, I don’t think it’s worth introducing it.

**Comment by CippoX** [comment]
@ahoppen seems reasonable, what about this last attempt? Otherwise, I'll just close the PR

**Comment by ahoppen** [comment]
This makes implicit assumptions about the number of children of a for loop that we are not interested in, which might change  if we changed the structure of for loops. I’d just close the PR, sorry.

**Pull request closed without merging**